### DISCUSSION TITLE
How to configure the Formatter?

### DISCUSSION THREAD

**Discussion by JoaquimEsteves** [Pure-D/serve-d#357] [Q&A] [answered]
Hello,

I'm brand new to the D lang - first of all I'd like to extends my thanks for providing us with such lovely free software.

My question revolves around the formatter that comes packaged with this LSP.
I can't seem to edit my preferred settings (for example: curly braces style)

I'm using NeoVim and have just installed this LSP, with the following (default) setup:

```lua
require'lspconfig'.serve_d.setup{}
```

I also made sure to add both `dcd-server` and `dcd-client` to my path.

I saw that the [dftm](https://github.com/dlang-community/dfmt?tab=readme-ov-file#dfmt-specific-properties) allows me to configure such things - but can't figure out the way to pass such information over through the LSP.

Is there some sort of config option I can set on `dub.json`?

Furthermore - is there I way I can set it up without having a `dub.json`? I'd like for my formatting options to be constant even when I'm just using writing to some one off file (for example `rdmd my_cool_script.d`)

**Comment by WebFreak001** [comment]
this depends on your LSP client and how you can pass configuration to it. Basically serve-d takes the project configuration and in there there are options to configure with dfmt.

The available settings as D struct: https://github.com/Pure-D/serve-d/blob/1a26dcda75950531ff5a9c08a87993c05a51fa16/source/served/types.d#L119-L135

or as JSON schema: https://github.com/Pure-D/code-d/blob/b19258b82d40b72ac4e60497214de14a4248d4d8/package.json#L763

Line endings, indentation size and style (tabs/spaces) should be synced with your standard editor configuration. So if the editor inserts tabs/spaces, the auto-formatting should use the same. (otherwise it might be a NeoVim LSP bug)

You will have to consult the NeoVim LSP documentation to see how to set the configuration values. If you find a way, feel free to submit a PR to the [`editor-vim.md`](https://github.com/Pure-D/serve-d/blob/master/editor-vim.md) file in the repository

**Reply by JoaquimEsteves** [reply, marked as answer]
Ahhh! I see.

With your suggestions I was able to solve the braces issue like so:

```lua
lspconfig.serve_d.setup({
	on_attach = my_on_attach_func,
	settings = {
		dfmt = {
			braceStyle = "stroustrup", -- or your desired style 
		},
	},
})
```

I'd be glad to make a PR to the editor-vim.md; I just need a few pointers.

My nitpick is discoverability - if I type `DFmt` as the [types](https://github.com/Pure-D/serve-d/blob/1a26dcda75950531ff5a9c08a87993c05a51fa16/source/served/types.d#L119-L135) file suggests then it doesn't work - it has to be written in the same style as the `package.json`

That's fair enough, but it feels weird to me to tell the reader "read the package.json of a separate git repo to see how you can tweak serve-d".

I see that that's how it's setup in the [main README](https://github.com/Pure-D/serve-d/?tab=readme-ov-file#user-configuration) but I was wondering if there was a better way to point new users like myself to this information.

**Reply by WebFreak001** [reply]
yeah could add the information here as well - it would need to be automatically generated though since it's a pain to maintain multiple versions of the same thing. Could maybe add it to the output of `serve-d --help`

**Reply by JoaquimEsteves** [reply]
That would be absolutely ideal. Thanks so much for your help.
